[blog]: Gitpod installer blog post. (#1849)
gitpod-io/website: src/routes/blog/gitpod-installer.md

---
author: csweichel, MrSimonEmms
date: Wed, 6 April 2022 16:00:00 UTC
excerpt: We at Gitpod believe in making developer experiences better, and we are super excited to introduce Gitpod Installer 🎉, which makes Gitpod installations a breeze!
image: teaser.png
slug: gitpod-installer
subtitle:
teaserImage: teaser.png
title: Why we moved from Helm to Gitpod Installer
---

<script context="module">
  export const prerender = true;
</script>

We at Gitpod believe in making developer experiences better, and we are super excited to introduce Gitpod Installer 🎉, which makes Gitpod installations a breeze!

Gitpod is available as a SaaS and self-hosted product. It is composed of many components working together in a Kubernetes cluster. You can install everything in a single cluster or place separate components in multiple clusters. Initially, we used Helm to orchestrate the installation, but our Helm charts became complicated over time. Hence, the new Gitpod Installer came into existence.

## The fault with our Helm charts

> In the last couple of years, as we kept building more and more features for Gitpod, our Helm charts started accumulating a lot of logic with an ever-expanding configuration surface. It all quickly grew out of hand. The Helm charts "values.yaml" rose to 750 lines. Many of these were not configurations but necessities to workaround due to Helm limitations.

Reviewer: ```suggestion
In the last couple of years, as we kept building more and more features for Gitpod, our Helm charts started accumulating a lot of logic with an ever-expanding configuration surface. It all quickly grew out of hand. The Helm charts `values.yaml` rose to 750 lines. Many of these were not configurations but necessities to workaround due to Helm limitations.
```

> To make the matter worse, we had many undocumented configuration values that we used only in production or core-dev but did not make sense for any other environment. The entirely of the Helm charts grew into a mammoth 7000 lines to accommodate all the variations. Helm charts are made from templated YAML using Go templates. At this scale, it was a tough job to maintain them. It became easy to make mistakes such as missing whitespaces and realising it much later while deploying and then spending hours spotting the root issue. Lack of editor support and static checks for Helm charts forced you to refer to the documentation often. All of this made it tedious to write Kubernetes configurations. It sometimes led to config drift between what your application expects and what Helm generated, leading to production outages.

Reviewer: ```suggestion
To make the matter worse, we had many undocumented configuration values that we used only in production or core-dev but did not make sense for any other environment. The entirety of the Helm charts grew into a mammoth 7000 lines to accommodate all the variations. Helm charts are made from templated YAML using Go templates. At this scale, it was a tough job to maintain them. It became easy to make mistakes such as missing whitespaces and realising it much later while deploying and then spending hours spotting the root issue. Lack of editor support and static checks for Helm charts forced you to refer to the documentation often. All of this made it tedious to write Kubernetes configurations. It sometimes led to config drift between what your application expects and what Helm generated, leading to production outages.
```

## Gitpod Installer, the saviour

> We started building the [Gitpod Installer](https://github.com/gitpod-io/gitpod/tree/main/install/installer) to make installation of Gitpod and maintenance of the installation scripts less painful. We wrote Gitpod Installer in Go, making it easier to manage, structure and reuse the code. Instead of dealing with YAML and templates, we deal with structs and functions which gets ustype safety, runtime checks, and composition.

Reviewer: ```suggestion
We started building the [Gitpod Installer](https://github.com/gitpod-io/gitpod/tree/main/install/installer) to make installation of Gitpod and maintenance of the installation scripts less painful. We wrote Gitpod Installer in Go, making it easier to manage, structure and reuse the code. Instead of dealing with YAML and templates, we deal with structs and functions which gets us type safety, runtime checks, and composition.
```

### Backwards compatibility to provide a clean update path

We chose to make the configuration file used by the Gitpod Installer carry an `apiVersion`. While loading the configuration, the Installer looks for the version and uses the corresponding loader. Each version contains its own set of validations, default settings and a migration path to higher versions. It allows us to introduce breaking changes in the configuration when needed while still providing compatibility and a migration path to configuration written for an older version.
Achieving this is extremely difficult with Helm charts. We had to provide additional logic for backward compatibility on a best effort basis.

#### Version Manifest

A Gitpod installation is a collection of builds of numerous projects. The version manifest provides the bill-of-materials required to run a complete Gitpod installation for a particular release. It guarantees that a specific release of Gitpod always runs with a deterministic set of services and dependencies. We generate the version manifest during build time and embed it into the Gitpod Installer.

#### Helm Dependency Integration

> Gitpod also depends on several third-party dependencies such as Minio, MySQL and RabbitMQ. Instead of reimplementing all the services to fit into the Installer framework, we reuse the community maintained Helm charts. We embed the respective Helm charts into the Installer, and during installation, we provide the required values and render the templates in isolation just like "helm template" would do. It helps us to leverage the excellent work already done by the Helm community while retaining control over the installation process.

Reviewer: ```suggestion
Gitpod also depends on several third-party dependencies such as Minio, MySQL and RabbitMQ. Instead of reimplementing all the services to fit into the Installer framework, we reuse the community maintained Helm charts. We embed the respective Helm charts into the Installer, and during installation, we provide the required values and render the templates in isolation just like `helm template` would do. It helps us to leverage the excellent work already done by the Helm community while retaining control over the installation process.
```

1. There can be only one installation per cluster which is due to
   - Fixed node affinities for services.
   - Fixed NodePorts for daemonsets.
2. Fix replica counts for services.
3. External helm dependencies are embedded, hence non-configurable. We chose to do this because it gives us confidence that Gitpod is always running against a dependency that we have validated (like a known MySQL/RabbitMQ version)
4. Cert-manager is a required dependency for creating internal SSL certificates for components like the Docker registry as it eliminates manual cert renewal.
5. The Installer requires external certificates to be provided as a Kubernetes secret (tls.key and tls.crt). By requiring just the secret to be present, it allows for users who want to user cert-manager to provide their certificates as well as those who want to provide their certificates via a specific certificate authority.

## Building Over Helm

> While building Gitpod, we have reused a lot of logic from Helm. We use the logic from "helm template render" to embed Helm charts for third-party dependencies and render them together with Installer components.

Reviewer: ```suggestion
While building Gitpod, we have reused a lot of logic from Helm. We use the logic from `helm template render` to embed Helm charts for third-party dependencies and render them together with Installer components.
```
